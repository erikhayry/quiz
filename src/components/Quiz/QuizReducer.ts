export enum ACTION_TYPE {
  Fetched = "fetched",
  Answer = "answer",
}

type ACTION =
  | { type: ACTION_TYPE.Fetched; questions: Question[] }
  | { type: ACTION_TYPE.Answer; answer: number; year: number };

export interface State {
  questions: Question[];
  answers: Record<number, Answer>;
}

export function reducer(state: State, action: ACTION): State {
  switch (action.type) {
    case ACTION_TYPE.Fetched: {
      return {
        ...state,
        questions: action.questions,
      };
    }
    case ACTION_TYPE.Answer: {
      const { year, answer } = action;
      const { answers } = state;

      answers[year] = {
        year,
        answer,
      };

      return { ...state, answers };
    }
  }
  return state;
}
