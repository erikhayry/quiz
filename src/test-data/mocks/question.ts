import { QUIZ_LENGTH } from "../../utils/config";

export function getMockQuestions(): Question[] {
  return new Array(QUIZ_LENGTH).fill("").map((_, index) => ({
    name: `Mock Winner ${index}`,
    alternatives: [2000 + index, 2000 + index + 1, 2000 + index + 2],
    year: 2000 + index,
  }));
}
