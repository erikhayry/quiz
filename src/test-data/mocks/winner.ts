import { QUIZ_LENGTH } from "../../utils/config";

export function getMockWinners(): Winner[] {
  return new Array(QUIZ_LENGTH).fill("").map((_, index) => ({
    name: `Mock Winner ${index}`,
    year: 2000 + index,
  }));
}
