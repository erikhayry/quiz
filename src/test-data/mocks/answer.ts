import { Answers } from "../../components/Quiz/QuizReducer";

export const MOCK_ANSWER_YEAR = 2000;
export const mockCorrectAnswer: Answer = {
  answer: MOCK_ANSWER_YEAR,
  year: MOCK_ANSWER_YEAR,
};
export const mockWrongAnswer: Answer = { answer: MOCK_ANSWER_YEAR, year: 2001 };
export const mockAnswers: Answers = {
  [MOCK_ANSWER_YEAR]: mockCorrectAnswer,
};
