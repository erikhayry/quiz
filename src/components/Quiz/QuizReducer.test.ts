import { ACTION_TYPE, reducer, State } from "./QuizReducer";
import { getMockQuestions } from "../../test-data/mocks/question";

const INITIAL_STATE: State = {
  questions: [],
  answers: {},
};

describe("Quiz reducer", () => {
  it("actions type: fetched", () => {
    const state = reducer(INITIAL_STATE, {
      type: ACTION_TYPE.Fetched,
      questions: getMockQuestions(),
    });

    expect(state).toMatchSnapshot();
  });

  it("actions type: answer", () => {
    const state = reducer(INITIAL_STATE, {
      type: ACTION_TYPE.Answer,
      year: 2020,
      answer: 2020,
    });

    expect(state.answers[2020]).toEqual({
      year: 2020,
      answer: 2020,
    });

    const updatedState = reducer(state, {
      type: ACTION_TYPE.Answer,
      year: 2020,
      answer: 2021,
    });

    expect(updatedState.answers[2020]).toEqual({
      year: 2020,
      answer: 2021,
    });
  });
});
