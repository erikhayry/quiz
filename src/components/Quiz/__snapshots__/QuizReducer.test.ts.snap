// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz reducer actions type: fetched 1`] = `
Object {
  "answers": Object {},
  "questions": Array [
    Object {
      "alternatives": Array [
        2000,
        2001,
        2002,
      ],
      "name": "Mock Winner 0",
      "year": 2000,
    },
    Object {
      "alternatives": Array [
        2001,
        2002,
        2003,
      ],
      "name": "Mock Winner 1",
      "year": 2001,
    },
    Object {
      "alternatives": Array [
        2002,
        2003,
        2004,
      ],
      "name": "Mock Winner 2",
      "year": 2002,
    },
    Object {
      "alternatives": Array [
        2003,
        2004,
        2005,
      ],
      "name": "Mock Winner 3",
      "year": 2003,
    },
    Object {
      "alternatives": Array [
        2004,
        2005,
        2006,
      ],
      "name": "Mock Winner 4",
      "year": 2004,
    },
    Object {
      "alternatives": Array [
        2005,
        2006,
        2007,
      ],
      "name": "Mock Winner 5",
      "year": 2005,
    },
    Object {
      "alternatives": Array [
        2006,
        2007,
        2008,
      ],
      "name": "Mock Winner 6",
      "year": 2006,
    },
    Object {
      "alternatives": Array [
        2007,
        2008,
        2009,
      ],
      "name": "Mock Winner 7",
      "year": 2007,
    },
    Object {
      "alternatives": Array [
        2008,
        2009,
        2010,
      ],
      "name": "Mock Winner 8",
      "year": 2008,
    },
    Object {
      "alternatives": Array [
        2009,
        2010,
        2011,
      ],
      "name": "Mock Winner 9",
      "year": 2009,
    },
  ],
}
`;
